pageserver/src/tenant/timeline/layer_manager.rs

```diff
@@ -0,0 +1,310 @@
+            self.layer_map.open_layer = None;
+            self.layer_map.next_open_layer_at = Some(end_lsn);
+            true
+        } else {
+            false
+        }
+    }
+
+    /// Pop the last frozen layer from the layer map, called from `flush_frozen_layer`.
+    pub fn flush_frozen_layer(&mut self) -> Option<Arc<InMemoryLayer>> {
+        self.layer_map.frozen_layers.pop_front()
+    }
+
+    /// Add image layers to the layer map, called from `create_image_layers`.
+    pub fn create_image_layers(&mut self, image_layers: Vec<ImageLayer>) {
+        let mut updates = self.layer_map.batch_update();
+        for layer in image_layers {
+            Self::insert_historic_layer(Arc::new(layer), &mut updates, &mut self.layer_fmgr);
+        }
+        updates.flush();
+    }
+
+    /// Insert into the layer map when a new delta layer is created, called from `create_delta_layer`.
+    pub fn flush_l0_delta_layer(&mut self, delta_layer: Arc<DeltaLayer>) {
+        let mut updates = self.layer_map.batch_update();
+        Self::insert_historic_layer(delta_layer, &mut updates, &mut self.layer_fmgr);
+        updates.flush();
+    }
+
+    /// Called when compaction is completed.
+    pub fn compact_l0(
+        &mut self,
+        layer_removal_cs: Arc<tokio::sync::OwnedMutexGuard<()>>,
+        compact_from: Vec<Arc<dyn PersistentLayer>>,
+        compact_to: Vec<Arc<dyn PersistentLayer>>,
+        metrics: &TimelineMetrics,
+    ) -> Result<()> {
+        let mut updates = self.layer_map.batch_update();
+        for l in compact_to {
+            Self::insert_historic_layer(l, &mut updates, &mut self.layer_fmgr);
+        }
+        for l in compact_from {
+            // NB: the layer file identified by descriptor `l` is guaranteed to be present
+            // in the LayerFileManager because we kept holding `layer_removal_cs` the entire
+            // time, even though we dropped `Timeline::layers` inbetween.
+            Self::delete_historic_layer(
+                layer_removal_cs.clone(),
+                l,
+                &mut updates,
+                metrics,
+                &mut self.layer_fmgr,
+            )?;
+        }
+        updates.flush();
+        Ok(())
+    }
+
+    /// Called when garbage collect the timeline.
+    pub fn gc_timeline(
+        &mut self,
+        layer_removal_cs: Arc<tokio::sync::OwnedMutexGuard<()>>,
+        gc_layers: Vec<Arc<dyn PersistentLayer>>,
+        metrics: &TimelineMetrics,
+    ) -> Result<ApplyGcResultGuard> {
+        let mut updates = self.layer_map.batch_update();
+        for doomed_layer in gc_layers {
+            Self::delete_historic_layer(
+                layer_removal_cs.clone(),
+                doomed_layer,
+                &mut updates,
+                metrics,
+                &mut self.layer_fmgr,
+            )?; // FIXME: schedule succeeded deletions in timeline.rs `gc_timeline` instead of in batch?
+        }
+        Ok(ApplyGcResultGuard(updates))
+    }
```

Comment:
First: how about `finish_gc_timeline` ?

---

That FIXME is a copy-over from original timeline.rs code

```rust
  let mut layer_names_to_delete = Vec::with_capacity(layers_to_remove.len());
            {
                for doomed_layer in layers_to_remove {
                    layer_names_to_delete.push(doomed_layer.filename());
                    self.delete_historic_layer(
                        layer_removal_cs.clone(),
                        doomed_layer,
                        &mut updates,
                        mapping,
                    )?; // FIXME: schedule succeeded deletions before returning?
                    result.layers_removed += 1;
                }
            }

            if result.layers_removed != 0 {
                fail_point!("after-timeline-gc-removed-layers");
            }

            if let Some(remote_client) = &self.remote_client {
                remote_client.schedule_layer_file_deletion(&layer_names_to_delete)?;
            }
```

That got me thinking: why aren't the `remote_client` interactions (schedule of creation, deletion) part of the `LayerManager`?

After all, `LayerManager`s job is to keep things coherent.

I find the `ApplyGcResultGuard` quite confusing at the call site in `timeline.rs`.
At a minimum, add a comment to this function here, explaining what the guard does.

But, we should really think about how `remote_client` interactions relate to `LayerManager`. Let's do it in this thread.

Reply:
I think at some time we should put all of them in one place. A sketch in my mind:

* all layer operations should go through layer manager
* in each of the functions we have in this PR, apart from modifying layer map, we should also append to MANIFEST
* there will be a remote sync thread that subscribes to MANIFEST updates, receive them in the order they are applied locally, and apply them to the remote storage in the background.

But I realized all these things after my refactor batch. I didn't consider remote operations when I was refactoring it. Probably it would be good to move more things into layer manager in future PRs.

Also one thing I noticed is that the operation on layer map and the operation on remote layer manager is not always in-sync. Somethings we do remote operations before layer map operations (in the case of compaction), and some other times we do it after layer map operations (when GC / reconcile with remote). So I guess moving remote operations into layer manager is harder to implement than the current PR, but it looks like a good direction to go.

`ApplyGcResultGuard` is here to keep all operation sequence *exactly* the same as before because previously we apply the layer map flush after the remote layer operations. I have thought about directly doing all the flushes inside layer manager, but there might be some race condition things that I'm not aware of, and therefore I did this trick to make everything executes in the exact order as before.

Reply:
... actually as long as we hold the `guard` write lock in `timeline.rs` it should not matter where to flush as no one else should be able to access the layer map? This way we can simply remove the guard and flush inside the layer manager function. I was thinking about the case that remote upload fails, but even in that case it does not matter where we do the manual flush...